File: nodejs/lib/parallel.js
'use strict';

const { Worker, isMainThread, parentPort } = require('worker_threads');
const os = require('os');
const path = require('path');

// ── Worker code (isMainThread === false) ──────────────────────
if (!isMainThread) {
  const fs = require('fs');
  const { Packer } = require('./packer');
  const { Unpacker } = require('./unpacker');

  parentPort.on('message', (msg) => {
    try {
      const startTime = Date.now();

      if (msg.type === 'pack') {
        const input = fs.readFileSync(msg.inputPath);
        const packer = new Packer(msg.opts || {});
        const output = packer.pack(input);
        fs.writeFileSync(msg.outputPath, output);
        parentPort.postMessage({
          type: 'result',
          taskId: msg.taskId,
          success: true,
          stats: {
            inputSize: input.length,
            outputSize: output.length,
            elapsed: Date.now() - startTime,
            dictSize: packer.dictChunks.length,
          },
        });
      } else if (msg.type === 'unpack') {
        const input = fs.readFileSync(msg.inputPath);
        const unpacker = new Unpacker();
        const output = unpacker.unpack(input);
        fs.writeFileSync(msg.outputPath, output);
        parentPort.postMessage({
          type: 'result',
          taskId: msg.taskId,
          success: true,
          stats: {
            inputSize: input.length,
            outputSize: output.length,
            elapsed: Date.now() - startTime,
            dictSize: unpacker.dict.length,
          },
        });
      } else {
        parentPort.postMessage({
          type: 'result',
          taskId: msg.taskId,
          success: false,
          error: `Unknown task type: ${msg.type}`,
        });
      }
    } catch (err) {
      parentPort.postMessage({
        type: 'result',
        taskId: msg.taskId || 'unknown',
        success: false,
        error: err.message,
      });
    }
  });

  // Worker ready signal
  parentPort.postMessage({ type: 'ready' });
}

// ── Main process code ──────────────────────────────────────

/**
 * Worker Thread based parallel processing pool
 *
 * Usage:
 *   const pool = new WorkerPool(4);
 *   pool.init();
 *   const results = await pool.runAll(tasks);
 *   await pool.destroy();
 */
class WorkerPool {
  /**
   * @param {number} [size] - Number of workers (default: CPU core count)
   */
  constructor(size) {
    this._size = size || os.cpus().length;
    this._workers = [];
    this._idle = [];         // Idle worker queue
    this._taskQueue = [];    // Pending task queue
    this._taskIdCounter = 0;
    this._pendingTasks = new Map(); // taskId → { resolve, reject }
    this._initialized = false;
  }

  /**
   * Initialize worker pool
   */
  init() {
    if (this._initialized) return;
    this._initialized = true;

    for (let i = 0; i < this._size; i++) {
      const worker = new Worker(__filename);
      worker.on('message', (msg) => this._onMessage(worker, msg));
      worker.on('error', (err) => this._onError(worker, err));
      this._workers.push(worker);
      this._idle.push(worker);
    }
  }

  /**
   * Handle worker message
   */
  _onMessage(worker, msg) {
    if (msg.type === 'ready') return;

    if (msg.type === 'result') {
      const pending = this._pendingTasks.get(msg.taskId);
      if (pending) {
        this._pendingTasks.delete(msg.taskId);
        if (msg.success) {
          pending.resolve({ success: true, stats: msg.stats });
        } else {
          pending.resolve({ success: false, error: msg.error });
        }
      }

      // Assign pending tasks from queue if available
      if (this._taskQueue.length > 0) {
        const next = this._taskQueue.shift();
        worker.postMessage(next.message);
      } else {
        this._idle.push(worker);
      }
    }
  }

  /**
   * Handle worker error
   */
  _onError(worker, err) {
    // Fail all pending tasks for this worker
    for (const [taskId, pending] of this._pendingTasks) {
      pending.resolve({ success: false, error: err.message });
      this._pendingTasks.delete(taskId);
    }
  }

  /**
   * Execute single task
   * @param {Object} task - { type, inputPath, outputPath, opts }
   * @returns {Promise<Object>} { success, stats } | { success, error }
   */
  runTask(task) {
    const taskId = String(++this._taskIdCounter);
    const message = { ...task, taskId };

    return new Promise((resolve, reject) => {
      this._pendingTasks.set(taskId, { resolve, reject });

      if (this._idle.length > 0) {
        const worker = this._idle.shift();
        worker.postMessage(message);
      } else {
        this._taskQueue.push({ message });
      }
    });
  }

  /**
   * Execute multiple tasks concurrently
   * @param {Array<Object>} tasks
   * @returns {Promise<Array<Object>>}
   */
  async runAll(tasks) {
    const promises = tasks.map(task => this.runTask(task));
    return Promise.all(promises);
  }

  /**
   * Destroy pool
   */
  async destroy() {
    const terminations = this._workers.map(w => w.terminate());
    await Promise.all(terminations);
    this._workers = [];
    this._idle = [];
    this._initialized = false;
  }
}

// Export only from main process
if (isMainThread) {
  module.exports = { WorkerPool };
}

File: python/src/mshzip/parallel.py
'ProcessPoolExecutor based parallel processing'
from __future__ import annotations

import os
import time
from concurrent.futures import ProcessPoolExecutor
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from .packer import Packer
from .unpacker import Unpacker
from .constants import DEFAULT_CHUNK_SIZE, DEFAULT_FRAME_LIMIT, DEFAULT_CODEC, DEFAULT_SUB_CHUNK_SIZE


@dataclass
class TaskResult:
    'Task result.'
    success: bool
    input_size: int = 0
    output_size: int = 0
    elapsed_ms: float = 0.0
    dict_size: int = 0
    error: str = ''


@dataclass
class Task:
    'Parallel task definition.'
    type: Literal['pack', 'unpack']
    input_path: str
    output_path: str
    chunk_size: int | str = DEFAULT_CHUNK_SIZE
    frame_limit: int = DEFAULT_FRAME_LIMIT
    codec: str = DEFAULT_CODEC
    crc: bool = False
    hier_dedup: str | bool = 'auto'
    sub_chunk_size: int = DEFAULT_SUB_CHUNK_SIZE


def _worker_fn(task: Task) -> TaskResult:
    'Function executed in worker process.'
    try:
        start = time.monotonic()
        input_data = Path(task.input_path).read_bytes()

        if task.type == 'pack':
            packer = Packer(
                chunk_size=task.chunk_size,
                frame_limit=task.frame_limit,
                codec=task.codec,
                crc=task.crc,
                hier_dedup=task.hier_dedup,
                sub_chunk_size=task.sub_chunk_size,
            )
            output_data = packer.pack(input_data)
            Path(task.output_path).write_bytes(output_data)
            elapsed = (time.monotonic() - start) * 1000
            return TaskResult(
                success=True,
                input_size=len(input_data),
                output_size=len(output_data),
                elapsed_ms=elapsed,
                dict_size=len(packer.dict_chunks),
            )
        elif task.type == 'unpack':
            unpacker = Unpacker()
            output_data = unpacker.unpack(input_data)
            Path(task.output_path).write_bytes(output_data)
            elapsed = (time.monotonic() - start) * 1000
            return TaskResult(
                success=True,
                input_size=len(input_data),
                output_size=len(output_data),
                elapsed_ms=elapsed,
                dict_size=len(unpacker.dict),
            )
        else:
            return TaskResult(
                success=False,
                error=f'Unknown task type: {task.type}',
            )
    except Exception as e:
        return TaskResult(success=False, error=str(e))


class WorkerPool:
    'ProcessPoolExecutor based parallel processing pool.'

    def __init__(self, size: int | None = None) -> None:
        self._size = size or os.cpu_count() or 4
        self._executor: ProcessPoolExecutor | None = None

    def init(self) -> None:
        'Initialize pool.'
        if self._executor is None:
            self._executor = ProcessPoolExecutor(max_workers=self._size)

    def run_task(self, task: Task) -> TaskResult:
        'Run single task.'
        if self._executor is None:
            self.init()
        assert self._executor is not None
        future = self._executor.submit(_worker_fn, task)
        return future.result()

    def run_all(self, tasks: list[Task]) -> list[TaskResult]:
        'Run multiple tasks concurrently.'
        if self._executor is None:
            self.init()
        assert self._executor is not None
        futures = [self._executor.submit(_worker_fn, t) for t in tasks]
        return [f.result() for f in futures]

    def shutdown(self) -> None:
        'Shutdown pool.'
        if self._executor:
            self._executor.shutdown(wait=True)
            self._executor = None
